iota/types.py

```diff
@@ -36,830 +35,829 @@
-    elif isinstance(other, text_type):
-      return TryteString(self._trytes + other.encode('ascii'))
-    elif isinstance(other, (binary_type, bytearray)):
-      return TryteString(self._trytes + other)
-    else:
-      raise with_context(
-        exc = TypeError(
-          'Invalid type for TryteString concatenation '
-          '(expected Union[TryteString, {binary_type}, bytearray], '
-          'actual {type}).'.format(
-            binary_type = binary_type.__name__,
-            type        = type(other).__name__,
-          ),
-        ),
-
-        context = {
-          'other': other,
-        },
-      )
-
-  def __eq__(self, other):
-    # type: (TrytesCompatible) -> bool
-    if isinstance(other, TryteString):
-      return self._trytes == other._trytes
-    elif isinstance(other, text_type):
-      return self._trytes == other.encode('ascii')
-    elif isinstance(other, (binary_type, bytearray)):
-      return self._trytes == other
-    else:
-      raise with_context(
-        exc = TypeError(
-          'Invalid type for TryteString comparison '
-          '(expected Union[TryteString, {binary_type}, bytearray], '
-          'actual {type}).'.format(
-            binary_type = binary_type.__name__,
-            type        = type(other).__name__,
-          ),
-        ),
-
-        context = {
-          'other': other,
-        },
-      )
-
-  # :bc: In Python 2 this must be defined explicitly.
-  def __ne__(self, other):
-    # type: (TrytesCompatible) -> bool
-    return not (self == other)
-
-  def count_chunks(self, chunk_size):
-    # type: (int) -> int
-    """
-    Returns the number of constant-size chunks the TryteString can be
-    divided into (rounded up).
-
-    :param chunk_size:
-      Number of trytes per chunk.
-    """
-    return len(self.iter_chunks(chunk_size))
+    def __init__(self, trytes, pad=None):
+        # type: (TrytesCompatible, Optional[int]) -> None
+        """
+        :param trytes:
+            Byte string or bytearray.
+
+        :param pad:
+            Ensure at least this many trytes.
+
+            If there are too few, null trytes will be appended to the
+            TryteString.
+
+            Note:    If the TryteString is too long, it will _not_ be
+            truncated!
+        """
+        super(TryteString, self).__init__()
+
+        if isinstance(trytes, (int, float)):
+            raise with_context(
+                exc=TypeError(
+                    'Converting {type} is not supported; '
+                    '{cls} is not a numeric type.'
+                    .format(
+                        type=type(trytes).__name__,
+                        cls=type(self).__name__
+                    )),
+                context={
+                    'trytes': trytes,
+                },
+            )
+
+        if isinstance(trytes, TryteString):
+            if type(trytes) is TryteString or issubclass(
+              type(trytes), type(self)):
```

Comment:
Could we restore `incoming_type`?  It might be unnecessary technically, but we don't need to refactor it for PEP-8 compliance.

This is already a huge PR that touches virtually every file in the codebase; I'd prefer it if we limited the scope of functional changes as much as possible.